Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp

```diff
@@ -219,39 +379,150 @@ void KafkaEventStreamDecoder::captureImplExcept() {
+                                                   size_t &eventCount,
+                                                   uint64_t &pulseTimeRet) {
+  EASY_FUNCTION();
+
+  /* Parse message */
+  const auto eventMsg =
       GetEventMessage(reinterpret_cast<const uint8_t *>(buffer.c_str()));
 
-  DateAndTime pulseTime = static_cast<int64_t>(eventMsg->pulse_time());
+  /* Parse pulse time */
+  pulseTimeRet = static_cast<uint64_t>(eventMsg->pulse_time());
+  const DateAndTime pulseTime(pulseTimeRet);
+
+  /* Get TOF and detector ID buffers */
   const auto &tofData = *(eventMsg->time_of_flight());
   const auto &detData = *(eventMsg->detector_id());
-  auto nEvents = tofData.size();
 
-  DataObjects::EventWorkspace_sptr periodBuffer;
-  std::lock_guard<std::mutex> lock(m_mutex);
+  /* Increment event count */
+  const auto nEvents = tofData.size();
+  EASY_VALUE("rx_event_count", nEvents);
+  eventCount += nEvents;
+
+  /* Create buffered pulse */
+  BufferedPulse pulse{pulseTime, 0};
+
+  /* Perform facility specific operations */
   if (eventMsg->facility_specific_data_type() == FacilityData_ISISData) {
-    auto ISISMsg =
+    std::lock_guard<std::mutex> workspaceLock(m_mutex);
+    const auto ISISMsg =
         static_cast<const ISISData *>(eventMsg->facility_specific_data());
-    periodBuffer = m_localEvents[static_cast<size_t>(ISISMsg->period_number())];
-    auto &mutableRunInfo = periodBuffer->mutableRun();
+    pulse.periodNumber = static_cast<int>(ISISMsg->period_number());
+    auto periodWs = m_localEvents[pulse.periodNumber];
+    auto &mutableRunInfo = periodWs->mutableRun();
     mutableRunInfo.getTimeSeriesProperty<double>(PROTON_CHARGE_PROPERTY)
         ->addValue(pulseTime, ISISMsg->proton_charge());
-  } else {
-    periodBuffer = m_localEvents[0];
   }
-  for (decltype(nEvents) i = 0; i < nEvents; ++i) {
-    auto &spectrum = periodBuffer->getSpectrum(
-        m_specToIdx[static_cast<int32_t>(detData[i])]);
-    spectrum.addEventQuickly(TofEvent(static_cast<double>(tofData[i]) *
-                                          1e-3, // nanoseconds to microseconds
-                                      pulseTime));
+
+  const auto starttime = std::chrono::system_clock::now();
+
+  {
+    std::lock_guard<std::mutex> bufferLock(m_intermediateBufferMutex);
+
+    /* Store the buffered pulse */
+    m_receivedPulseBuffer.push_back(pulse);
+    const auto pulseIndex = m_receivedPulseBuffer.size() - 1;
+
+    /* Ensure storage for newly received events */
+    const auto oldBufferSize(m_receivedEventBuffer.size());
+    m_receivedEventBuffer.reserve(oldBufferSize + nEvents);
+
+    /* Store the buffered events */
+    std::transform(
+        detData.begin(), detData.end(), detData.begin(),
+        std::back_inserter(m_receivedEventBuffer),
+        [&](uint64_t detId, uint64_t tof) -> BufferedEvent {
+          return {m_specToIdx[static_cast<int>(detId)], tof, pulseIndex};
```

Comment:
Have you checked how much time is spent in this look-up in `std::unordered_map`? For the different but similar case of loading files I had previously experimented with using a per-bank fixed integer offset instead. Assuming that all pixel IDs within a bank come without gaps you can then eliminate the map lookup and replace it by an integer addition.